### ISSUE TITLE
Automatic installation of git during GE install

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#4967]
Extracted from #4515

@bimimicah:

Would anyone be interested in code allowing git to be downloaded during the install process, rather than bundling it with the installer? This would allow us to keep helping users update git, with the added bonus of always installing the latest version.

According to the [WiX Documentation](http://wixtoolset.org/documentation/manual/v3/bundle/authoring_bundle_package_manifest.html), we can create an \<ExePackage> that downloads the latest version during setup (we can add an install option/condition, to keep the choice with the user):

            <ExePackage 
              Name="MyPackage.exe"
              DetectCondition="ExeDetectedVariable"
              DownloadUrl="http://example.com/?mypackage.exe"
              InstallCommand="/q /ACTION=Install"
              RepairCommand="/q ACTION=Repair /hideconsole"
              UninstallCommand="/q ACTION=Uninstall /hideconsole" 
              InstallCondition="NOT VersionNT64 AND VersionNT >= v5.1" />

The most important part is to set DownloadUrl to the "latest" link for git for windows.
This could be an issue, because GitHub doesn't have a way (that I've found) to link directly to the latest release binary. So, to resolve that, I created a quick, draft-version AWS lambda function*:

[https://wkl5nf5m2e.execute-api.us-west-2.amazonaws.com/download-git-windows](https://wkl5nf5m2e.execute-api.us-west-2.amazonaws.com/download-git-windows)

    //API gateway must be configured to add "redirectLocation" as the HTTP Location header
    //the XMLHttpRequest node.js library must be included
    var sys = require('util');
    var XMLHttpRequest = require("XMLHttpRequest").XMLHttpRequest;
    
    exports.handler = (event, context, callback) => {
        
        var downloadpath = "";
        
        var xml = new XMLHttpRequest();
        xml.open("GET","https://api.github.com/repos/git-for-windows/git/releases/latest",false);
        xml.send();
        var responseJSON = JSON.parse(xml.responseText);
        downloadpath = responseJSON.assets[0].browser_download_url; //first release asset is 32-bit git - this will break if they change the order
        
        const httpResponse = {
            lastCheck: Math.round(Date.now() / 3600000),
            redirectlocation: downloadpath
        }
        
        callback(null, httpResponse);
    };

*Note:  API gateway must be configured to add "redirectLocation" as the HTTP Location header, and the XMLHttpRequest node.js library must be included

I hope this is helpful. I would be willing to take some of my free time to work on this, if anyone is interested. My personal opinion is that bundling git (in some form) would be in the best interests of GitExtensions's users (staying up-to-date), and I hope this (or something similar) would be a workable compromise which makes things simpler for the developers too.

**Comment by RussKie** [comment]
@gerhardol 

Automatic download is a nice touch, but not necessarily better.
(To some extent, the change is done in the way it is done because it was a simple way forward that seem good enough. The knowledge of the Wix installer is low.)

Something like adding download to the installer was my first thought, maybe just to add a link in the installer.
However, there is a benefit in linking to a wiki page (as done with the related PR) to point to benefit/issues with certain versions. For instance 2.16.0 had issues with reset. Supposedly, some Git versions changed where utilities are installed which broke GE.

**Comment by RussKie** [comment]
@bimimicah

    Automatic download is a nice touch, but not necessarily better.
    (To some extent, the change is done in the way it is done because it was a simple way forward
    that seem good enough. The knowledge of the Wix installer is low.)

I'm offering to figure out that part of the WiX installer. On the other hand, there would be little point in my putting in any more effort toward this idea _if the developers do not want the idea to happen, or want something else_.
That's why I'm trying to gauge the interest level, so I know whether to spend time on this.

If we _did_ bring back the Git install page with the automatic download method (or a link, for that matter), then adding a wiki link would be a good idea, in my opinion.

**Comment by RussKie** [comment]
@vbjay 

I for one would love to have automatic download. One idea is we could have
a latest supported version of git on the repo.  The installer downloads the
file and reads it. It downloads that version the file points to on the git
scm site. As we support later and later versions we just update the git
version file in the repo.

**Comment by RussKie** [comment]
@gerhardol 

@bimimicah
Fully understand where you are aiming, I do not know exactly what I want myself. Maybe a "simple" automatic download is not the solution we want?
Personally, I would not use it as I never use the installer myself and use latest versions at home (to test) and the version at work is supposed to be fixed.

A plain "download latest" feature with a URL to the wiki is the simplest solution that I would support myself (it is @RussKie that is in charge). Maybe we should go with a better solution?
Should there be a recommended version or should we just ignore Git depreciated features and changes and go for the latest. If not latest, we need a list like @vbjay suggested. Should that be in the repo or wiki? What format?

**Comment by RussKie** [comment]
@bimimicah

Whether we end up going with the latest version or a specified "approved", "stable" version, my personal opinion is that we should continue to offer an automatic install of git (and also kdiff3/some diff tool, but that's a separate issue).
Believe it or not, there are people out there who don't know anything about git or version control; there are plenty of industries/fields that benefit from using version control that have nothing to do with programming.
I personally, several years ago, learned how git worked by downloading GitExtensions and clicking around (also reading documentation, but let's be honest - the "git-scm" docs are complex and confusing to the un-initiated).
At the time, git was difficult to set up on windows (as I remember it anyway) and GitExtensions offered an easy way to deal with that (it provided a stable version in the installer with sensible defaults, and then helped you set it up on first run).

**Comment by RussKie** [comment]
Say we have a dedicated branch in which we can publish a file (json?) with min and recommended git versions.

* Do we want to read this file from the installer?
* Do we want to read the same file from GE, when checking what version of git is installed on the user's computer?
* Say we have released vX.Y.Z which was (kind of) tested against the latest version of git at the time of release (git vA.B.C). Some months later a user decides to install vX.Y.Z, however there is a new version of git (vE.F.G) and it may have some breaking changes compared to vA.B.C. How do we want to handle this situation?

**Comment by bimimicah** [comment]
@RussKie 
I suppose the first questions would be

1. How often does Git introduce breaking changes? I had always assumed this would be very rare, as Git is used all over the world, but you guys would have a much better idea.

1. If a breaking change is introduced, is it in primary functionality (commit, push, fetch, etc.) or in something rarely used (bisect? not sure what that does)?

1. Is number 2 above even relevant? Maybe all breakages are equally bad?

My personal opinion (assuming primary functionality breakages are rare to non-existent) would be to always download the latest Git upon installation, and rely on GitExtensions' update feature to update GE when there is a breakage. (Maybe add an update message to the updater, to clarify level of importance?) That would get users onto the latest versions of both the fastest, I think.

@gerhardol 
Of course, as I said earlier, if the user doesn't check "install Git" in setup, then it's up to them what version of Git they will need to install.

**Comment by vbjay** [comment]
I frankly think latest would be fine but I still think we need the insurance for the future.  I think a json file with a latest value until there is a breakage and then we change it to latest working version number.  When we have updated GE, we can revert it to say latest again.  In GE we could read the same value and alert the user to the breakage if they have a higher version.  

If the latest version caused a breakage we could alter the file from the first one to the second.  So if latest was 2.17.1.8 we could alert the user that they have a version we do not support because of ...  New installs would not install the breaking version.  Once we update GE to update GE, we could alter the file and GE and the installer would install the latest version and not alert on a breakage any more.  

```json
{supportedGitVersion:"latest"}
```

```json
{supportedGitVersion:"2.16.*"}
```

**Comment by gerhardol** [comment]
> I suppose the first questions would be

1a. Git changed where commands are installed in some releases, which broke GE some years ago I was told when asking that the installer should be removed. No first hand information.
I recall that the version included with GE was kept the same for a long time due to the installer to.

1b. Depreciation, adding of new output in non porcelain commands etc could _potentially_ break GE.
It is more common that options break GE though.
I am not ware of any Git version changes that breaks GE myself, but I have not been involved deeply in GE for so long. @vbjay ?

2. The core commands use porcelain output, that should not change. Especially options can cause functionality to break, for instance presenting diff for submodules.
It is something to design for at least, so what vbay said.

Note: GE checks the Git version when starting up, so the lowest version check in the installer is not so critical. The recommended version couldbe implemented in the core program too.

**Comment by vbjay** [comment]
My reasoning on the installer is to prevent new setups that don't have git installed don't install get + git in a broken state. If we try and it tasks. We update the json file which will present the issue till we are fixed. The GE check will be for the user installed a newer version and we can tell them there are issues and they should install x version till we support x version.

**Comment by bimimicah** [comment]
I think I understand the requirements better now, thank you.
I'm currently supporting @vbjay 's suggestion:

    {supportedGitVersion:"latest"}
    {supportedGitVersion:"2.16.*"}

I'm thinking we can 

1. grab the supported release value from the GitExtensions repository
1. call the GitHub releases api for the gitforwindows repository with that value
1. get the download url from that, stuff it in an MSI variable
1. which can be used in the &lt;downloadUrl> attribute of an &lt;exePackage>
1. let Windows Installer handle the rest of it.

I'll be busy with a project for the next several days, but I'll try to work on it when I get a chance, and see how feasible that workflow is...

**Comment by vbjay** [comment]
You got it.

**Comment by gerhardol** [comment]
My opinion it that there is a minimum version and a recommended version, that depends on the version to install (how it is implemented in GE core now). Maybe bad versions too - but probably skip that now. So the json (or whatever) will be more complicated.

Whatever is added to the installer should be implemented in GE core too.

**Comment by vbjay** [comment]
I agree. It should be the same. I don't see why we can't have a git version
check library that implements the logic and the installer uses too.

**Comment by RussKie** [comment]
To sum up:
- we would like to have a unified git version check (minimum and
recommended)
- both the installer and the GE will use the check

The check can be implemented as a PowerShell script.

One more thing - the check must handle offline state gracefully. E.g. if an
installer is copied to a machine with no direct Internet connection, the
installer should simply advise a user to install the latest version of git
available, if git is not detected.
The same goes for GE - if no Internet connection but git is available - it
should just work.

**Comment by vbjay** [comment]
Can powershell be used in the installer?  I know it can be done in GE.

**Comment by RussKie** [comment]
I would think that it can, don't know how though

**Comment by KindDragon** [comment]
Hi. Maybe this has already been discussed, but I think it's better if GitExt settings dialog (not the installer) was able to download MinGit instead of system Git. This will also be useful for the portable version of GitExt.

**Comment by vbjay** [comment]
Great idea. They have a way to build a portable installer of it already. I
played with it in trying to make a tools think drive for students. It
worked pretty well.

**Comment by EbenZhang** [comment]
I would not struggle with the installer. i think it is equivalent to check the min ver required, download and install when GE starts.

**Comment by KindDragon** [comment]
> In 2.5x the settings page has a "Download MSYSGIT" link to the "Git for Windows" download page.

Check how Source Tree does this. I'm thinking about similar behavior.

**Comment by RussKie** [comment]
any screenshots to share?

**Comment by bimimicah** [comment]
@gerhardol @KindDragon 
It would not be very difficult to re-use the same code to download:

- the Git installer (if the user chooses) for the GitExtensions installer
- Git portable/Git installer (user chooses one) from GitExtensions' main program

The GitHub releases API makes this very easy, since _the Git for Windows project already releases the portable version with the installer simultaneously_ in the same release bundle. The code function can be called with a parameter set to "installer|portable" and a parameter set to "win32|win64".

@EbenZhang 
I'm offering to do the work on the installer. It will be either later this week or next week that I can get to this, however.

**Comment by bimimicah** [comment]
@gerhardol 
It should not be too difficult to have the code that checks for the existence of git additionally check that it is the minimum version/not a bad version.

at the moment the JSON I'm thinking would look like this:

    {
        supportedGitVersion: "latest",    (or "2.16.3", etc. if the latest is broken)
        minimumGitVersion: "2.01.04",    (for example)
        badGitVersions: ["2.14.05","2.15.01"]
    }

Installer algorithm would look like this:

1. User checks the "Install Git" box in the installer
1. download our _gitversions.json_
1. call the new getGitforWindowsDownloadURL() function (or whatever it will end up being called) with the value of _supportedGitVersion_, the install type (installer), and the platform (win32|win64) -- gets the URL from GitHub releases
1. Windows Installer downloads and installs the file at the specified URL 

GitExtensions program algorithm would look like this:

1. download our _gitversions.json_
1. We check for existing Git (we already have code for this)
1. If no Git, goto **7**; have Git, get version
1. If git version is less than _minimumGitVersion_, goto **7**
1. If git version equals one of _badGitVersion_, goto **7**
1. **Stop** the algorithm, no need to install or update
1. Display the "update/install Git" button in settings page
1. User chooses the "Install Git" button in the settings page
1. User chooses either portable or installed Git 
1. call the new getGitforWindowsDownloadURL() function (or whatever it will end up being called) with the value of _supportedGitVersion_, the install type (portable|installer), and the platform (win32|win64) -- gets the URL from GitHub releases
1. If install type is **installer**, GE downloads and installs the file at the specified URL 
1. If install type is **portable**, 
    1. GE shellexecute()s the URL in the user's default browser - they have to figure out how to set it up... 
    1. or we could unzip it to a subdirectory of GE? **this would never work with UAC without a lot of work...**

**Comment by vbjay** [comment]
Looks good to me.  The portable extract could be in CommonApplicationData
or LocalApplicationData

CommonApplicationData  is my suggestion.
https://msdn.microsoft.com/en-us/library/system.environment.specialfolder(v=vs.110).aspx

**Comment by gerhardol** [comment]
@RussKie 
SourceTree
![image](https://user-images.githubusercontent.com/6248932/40387135-17eb67be-5e0c-11e8-936f-ec82ff3c9026.png)
Downloads from Atlassian
![image](https://user-images.githubusercontent.com/6248932/40387224-56e9ffc0-5e0c-11e8-931f-aa58ab506ceb.png)

@bimimicah 
Maybe a little more discussions than expected, thanks for wanting to improve GE.

For the algorithm:
6a. If supported version is "latest", get the version from GitHub, stop if version matches
6b. If installed is not same as supported, go to 7

12 ii Install by default to the user local profile (see vbay suggestion)

**Comment by bimimicah** [comment]
@vbjay @gerhardol 
_CommonApplicationData_ or _LocalApplicationData_ sounds like a good idea for 12.ii.
The main thought I have about that is **we should probably call it "embedded"** (like SourceTree) rather than "portable", as my understanding of "portable" means it is self-contained and does not substantially modify the user's system. (e.g. a copy that completely runs from a thumb-drive)
I would lean toward using _CommonApplicationData_ (install only once, instead of requiring each user to install their own), but I don't especially care either way.

@gerhardol 
I use GitExtensions a lot, for many different things, including at work, so I am interested in making sure it sticks around for a long time. Thank you all for being open to letting me contribute.

For point 6. above, the idea behind the way I planned the algorithm is, if the version the user has already installed is above the minimum and isn't a "known bad", then we don't bother the user about updating it.

To put it another way, the only times GitExtensions would say "you need to update Git" are:
- no Git at all
- Git too old
- bad Git verison

That means that, at least the way I was thinking, we wouldn't need to compare the latest with the current, _except for the installer_ (I missed that part).

I would assume that would be your desired outcome, since you mentioned that at home and at work you need/want to manually manage the Git installation. I may be misunderstanding though. We could obviously do it differently, I'm just explaining my thought process.

**Comment by vbjay** [comment]
So we do have a portable config.  It's not completely perfect but this will
help in that regard.  The settings are portable.  Relative path support for
when the IsPortable is true would be helpful.


Open Gitextensions.exe.config

**Comment by RussKie** [comment]
@bimimicah: top job with the algorithm :+1: 
It is missing sad paths though, such as:
* network is not available (offline installation), 
* failed to download gitversions.json or git install (e.g. proxy in the way, connection lost etc)
* failed to complete git installation

**Comment by EbenZhang** [comment]
if u put the gitversion.json as part of the installer then no need to download it. also one version of this file is specific to only couple of versions of  GE, not all, when min ver changed, u probably dont want the previous ver of GE to to download new git.

Just from my experience many many years ago, wix programming was tricky. not sure how is it now. i would personally not touch it unless really necessary.

**Comment by vbjay** [comment]
I don't think you get the point. The point is to be responsive to git
changes. We download the current jsin so that if someone runs an installer
a month later, it still is current.

**Comment by EbenZhang** [comment]
will  it be possible the installer downloaded month ago doesn't compatible with the new min ver of git?

**Comment by vbjay** [comment]
We could have found an issue. The whole point of the json is so we can
react to that issue. Recrwad this issue.

**Comment by EbenZhang** [comment]
Might be an edge case since git has pretty well backwards compatibility. Say GE2 compatible with git [2.0, 3.0], GENext compatible with git [2.5, 3.0], if I’m not wrong the json will have [2.5, 3.0], what happens if we actually found GE2 is just compatible with git [2.0, 2.4] after GE2 released a couple of months and the users have installed git 2.5 as recommended, are we going to tell the user that GE2 is not compatible with git 2.5, you need to use the next version of GE which will be released in hmm maybe 6 months?

**Comment by vbjay** [comment]
What do you have against future proofing?

**Comment by EbenZhang** [comment]
I don't understand why saying that there should be a range of compatible versions of git for a specific version of GE is somehow against future proofing.

**Comment by vbjay** [comment]
It's not a range.  Latest will support future versions.  If git releases a
version that breaks stuff, we change latest to the latest working version.
When we update GE to support the new version, we change back to latest.

**Comment by vbjay** [comment]
Good idea.

**Comment by bimimicah** [comment]
@vbjay 

    Just a thought. Use the same style of warning we use for in the middle of a rebase to say we are using an unsupported version of git in the browse form. That way, it's not hidden behind settings. They click on the warning and it takes them to the right spot in settings. Only show this if the warning is red.

I think that's a good idea, but it is probably outside the scope of what I'm suggesting here. What I'm offering to work on is mostly integrated into existing workflows.

**Comment by vbjay** [comment]
Sure.  This was just more an idea for after this gets pulled in.  Just a
bit more in your face warning than having to go to settings to see it.

**Comment by vbjay** [comment]
Exactly why I suggested the json file in the repo.

**Comment by RussKie** [comment]
I don't think we should overly concern ourselves with v2. We will sunset it once v3 is released. v2 will stay around for Linux folk, but I don't see us making any changes to it.
If* in the future we decide to release v4, we will sunset v3 once the new version is out.

--

* - because we are not in a business of tricking our users to shell coins for new versions, I don't expect v4 anytime soon

**Comment by KindDragon** [comment]
> download our gitversions.json

gitversions.json should depend on GitExt version. For example, in past they changed in one version of Git how they store git submodules, so we added support for this new submodules format only to new version of GitExt.

Why not just store  gitversions.json in the installer?

**Comment by KindDragon** [comment]
> We could include a minimumGitExtensionsVersion entry in the gitversions.json file.

We often not test old GitExt versions with latest Git, so we might forget to update that field

**Comment by bimimicah** [comment]
I apologize everyone, I do plan to finish this.
Some life and work things came up, so it will probably not be until at least August at this point.

**Comment by RussKie** [comment]
Dusting off the thread.
Is this something we still wish to pursue?

**Comment by vbjay** [comment]
Yes.

**Comment by vbjay** [comment]
If I get the chocolatey setup for it I think it would be a simple Choco upgrade git -y

**Comment by RussKie** [comment]
That's why I asked. If we plan to pursue choco path, is this work required?

**Comment by bimimicah** [comment]
I apologize for not having any progress on this yet. I've been planning to work on this, but I've been very busy with work and some personal things.
If I find time to work on this soon, is this still something you all want? (my proposal, I mean - without dependency on chocolatey; we could always auto-detect if git was installed with chocolatey and use that if available?)

**Comment by vbjay** [comment]
You don't want a chocolatey run on each start.

My proposal is the Json file in the repo.

Our app checks versions against Jon and if it is a version that needs an
upgrade use chocolatey to do so.

**Comment by RussKie** [comment]
Suggest we move to #5481

**Comment by bimimicah** [comment]
What about everyone who doesn't use chocolatey? We could detect if they installed GitExtensions/git with chocolatey, and only use @vbjay's chocolatey management if they do. For the typical installer, we could use something like what was discussed previously in this issue.

**Comment by vbjay** [comment]
I can install chocolatey too. It's just a powershell script.

**Comment by RussKie** [comment]
I'd like to stay as simple as possible.

Here's how I see it:
* If a user uses choco - use it to install, choco will handle all necessary dependencies
* Alternatively, the user choose to download and install msi 
  - if git isn't installed, the user is greeted with an error - see #5481
  - alternatively, the user is directed to our [Wiki page](https://github.com/gitextensions/gitextensions/wiki/Git) that explains dependencies

The above my personal opinion. Happy to discuss alternatives.

**Comment by vbjay** [comment]
My thoughts are this.

We already have execute stuff code. I can try to run choco or find it. If
not installed run the chocolatey install script. Then use chocolatey to
install different packages.

P4merge, git, putty...

**Comment by RussKie** [comment]
I'd remove all that code. Makes the codebase more complicated

**Comment by vbjay** [comment]
It really isn't. We already have ps execution code. All I would do is
extend it by using a chocolatey manager.

**Comment by vbjay** [comment]
We would list only approved ge packages like kdiff and p4 merge. We would
also use it for the dependency like putty and git.

**Comment by bimimicah** [comment]
I realize I don't make these decisions, but I'll add my 2 cents anyway. :)

1) What about all the people without Powershell on their system (especially Windows 7 users and anyone who has disabled it)? I doubt they would be okay with installing the Windows Management Framework update just for GitExtensions.

2) We've always provided a choice in the past (the installer allowed you to either install git, kdiff3, etc. or not install and the user would be reminded to manually locate git). I would prefer that there be a path forward for those who would prefer not to have a package manager on their system, especially when installed with GitExtensions into protected directories (like program files).

3) Specifically what I mean in the 2nd point is: ideally have a system built-in. If not, at least make chocolatey an opt-in item, rather than a default install option.

**Comment by bimimicah** [comment]
I guess what I would like to see is:
(based on @RussKie's comment above)

- If a user uses chocolatey:
    - since it just wraps our MSI, run the MSI with the "install git" option unset
    - because choco will handle all necessary dependencies

- Or the user directly downloads and installs MSI (I would assume the most common use case):
    - installer detects if git is installed
    - if it is (minimum version), do nothing
    - if it is not, allow the user to check the "install git" option
    - if the user chose to install, then download the [latest or "latest approved"] git installer and run it

I don't see this as being particularly complicated. I will have some free time over the next week or so to work on this if you still want this functionality.

Adding install/upgrade code to GitExtensions itself is a separate goal (my opinion), and not what I was originally suggesting anyway.

**Comment by RussKie** [comment]
> Or the user directly downloads and installs MSI (I would assume it's the most common use case):
> 
> * installer detects if git is installed
> * if it is (minimum version), do nothing
> * if it is not, allow the user to check the "install git" option
> * if the user chose to install, then download the [latest or "latest approved"] git installer and run it

Or, if git not installed, a user gets prompted to install it.

> I don't see this as being particularly complicated. I will have some free time over the next week or so to work on this if you still want this functionality.

By all means, give it a go.

**Comment by vbjay** [comment]
With my solution portable is also handled and the program is resiliant.

**Comment by vbjay** [comment]
![screenshot_20181114-084936](https://user-images.githubusercontent.com/3629489/48486790-7ff61f80-e7ea-11e8-906a-1123dd2d5114.png)
Windows 7 sp1 is minimum os.

**Comment by GintasS** [comment]
> > Or the user directly downloads and installs MSI (I would assume it's the most common use case):
> > 
> > * installer detects if git is installed
> > * if it is (minimum version), do nothing
> > * if it is not, allow the user to check the "install git" option
> > * if the user chose to install, then download the [latest or "latest approved"] git installer and run it
> 
> Or, if git not installed, a user gets prompted to install it.
> 
> > I don't see this as being particularly complicated. I will have some free time over the next week or so to work on this if you still want this functionality.
> 
> By all means, give it a go.

Confirming this, so voting to close.
